merkle.go

```diff
@@ -112,24 +99,40 @@ func (t *Tree) Proof() ([][]byte, error) {
 	return t.proof, nil
 }
 
-func (t *Tree) addNode(n node) error {
+// calcParent returns the parent node of two child nodes.
+func (t *Tree) calcParent(lChild, rChild node) node {
+	return node{
+		value:        t.hash(lChild.value, rChild.value),
+		onProvenPath: lChild.onProvenPath || rChild.onProvenPath,
+	}
+}
+
+func (t *Tree) addLeaf(n node) error {
 	var parent, lChild, rChild node
 	l := t.baseLayer
 	var lastCachingError error
+
+	// Loop through the layers, starting from the base layer.
 	for {
+		// Writing the node to its layer cache, if applicable.
 		if l.cache != nil {
 			_, err := l.cache.Write(n.value)
 			if err != nil {
 				lastCachingError = errors.New("error while caching: " + err.Error())
 			}
 		}
+
+		// If no node is pending, then this node is a left sibling,
+		// pending for its right sibling before its parent can be calculated.
 		if l.parking.value == nil {
 			l.parking = n
 			break
 		} else {
+			// This node is a right sibling.
 			lChild, rChild = l.parking, n
 			parent = t.calcParent(lChild, rChild)
-			// A given node is required in the proof iff its parent is an ancestor of a leaf whose membership in the
+
+			// A given node is required in the proof if its parent is an ancestor of a leaf whose membership in the
```

Comment:
`iff` is not a typo 🙂 it's "if and only if", perhaps we can write the long version...